src/org/rascalmpl/util/maven/Artifact.java

```diff
@@ -130,70 +133,159 @@ private static boolean isVersionRange(String version) {
+                return false;
+            }
+            ResolveKey other = (ResolveKey) obj;
+
+            return groupId.equals(other.groupId) && artifactId.equals(other.artifactId)
+                && excludes.equals(other.excludes);
+        }
+    }
+
+
+    /**
+     * The state of the resolver. It contains the artifact and the transitive set of exclusions.
+     */
+    static private class ResolveState {
+        private final Artifact artifact;
+        private final Set<WithoutVersion> excludes;
+
+        public ResolveState(Artifact artifact, Set<WithoutVersion> excludes) {
+            this.artifact = artifact;
+            this.excludes = excludes;
+        }
+
+        public void calculateClassPath(Scope forScope, Set<ResolveKey> alreadyIncluded, Map<WithoutVersion, String> resolvedVersions, Map<WithoutVersion, SortedSet<String>> rangedDeps, ArrayList<Artifact> result, MavenParser parser) {
+            artifact.calculateClassPath(forScope, alreadyIncluded, resolvedVersions, rangedDeps, excludes, result, parser);
+        }
+    }
+
+    /**
+     * We will use the first version we encounter (breadth-first) and cache it.
+     * Note that caching is done modulo transitive exclusions.
+     * For ranged versions we just use the first suitable version for the first range we encounter.
+     */
+    private void calculateClassPath(Scope forScope, Set<ResolveKey> alreadySolved, Map<WithoutVersion, String> resolvedVersions, Map<WithoutVersion, SortedSet<String>> rangedDeps,
+        Set<WithoutVersion> excludes, ArrayList<Artifact> result, MavenParser parser) {
+        var nextLevel = new ArrayList<ResolveState>(dependencies.size());
         for (var d : dependencies) {
             var coordinate = d.getCoordinate();
-            var withoutVersion = coordinate.versionLess();
+
+            boolean alreadyResolved;
+
+            var versionLess = coordinate.versionLess();
 
             var version = coordinate.getVersion();
             if (isVersionRange(version)) {
                 try {
-                    version = ourResolver.findLatestMatchingVersion(coordinate.getGroupId(), coordinate.getArtifactId(), version);
-                    coordinate = new ArtifactCoordinate(coordinate.getGroupId(), coordinate.getArtifactId(), version, coordinate.getClassifier());
+                    version = ourResolver.findLatestMatchingVersion(coordinate.getGroupId(), coordinate.getArtifactId(),
+                        version);
+                    coordinate = new ArtifactCoordinate(coordinate.getGroupId(), coordinate.getArtifactId(), version,
+                        coordinate.getClassifier());
                 }
                 catch (UnresolvableModelException e) {
-                    messages = messages.append(Messages.error("Version range error: " + e.getMessage(), d.getPomLocation()));
+                    messages.append(Messages.error("Version range error: " + e.getMessage(), d.getPomLocation()));
```

Comment:
this is a bug, `IList` is immutable, so the result of `messages.append` is now lost.